sandbox/libs/authn/src/main/java/org/opensearch/authn/PermissionFactory.java

```diff
@@ -0,0 +1,93 @@
+ * formatting requirements and legacy permissions which are not checked for validity on creation.
+ */
+public class PermissionFactory {
+
+    public final static String[] INVALID_CHARACTERS = new String[] { ":", "" }; // This is a placeholder for what may want to be banned
+
+    // A placeholder for the different resources which a permission may grant a permission based on
+    public final static String[] QUALIFIED_PERMISSION_TYPES = new String[] { "cluster", "indices", "plugin", "extension" };
+
+    /**
+     * This function creates a standard permission instance. It includes checking that the permission that is being created
+     * is properly formatted.
+     */
+    public Permission createPermission(String permissionString) {
+
+        Permission newPermission = new Permission(permissionString);
+        permissionIsValidFormat(newPermission);
+        return newPermission;
+    }
+
+    /**
+     * This function creates a permission without checking that the permission string is valid.
+     */
+    public Permission createLegacyPermission(String permissionString) {
+        return new Permission(permissionString);
+    }
+
+    /**
+     * Check that the permission does not contain any forbidden strings.
+     * Assumes that the permission is formatted as resource.action
+     */
+
+    public void permissionIsValidFormat(Permission permission) {
+
+        // Check for illegal characters in any of the permission segments O(3n)
+        for (String character : INVALID_CHARACTERS) {
+            if (permission.permissionType.contains(character) || permission.action.contains(character)) {
+                throw new InvalidPermissionException(
+                    "The provided permission string for '"
+                        + permission.permissionString
+                        + "' is not valid. The permission type and action  may not include "
+                        + "the character ':' or be empty."
+                );
+            }
+        }
+
+        // Make sure the resource being acted on is one of the qualified permission types
+        if (!new ArrayList<String>(List.of(QUALIFIED_PERMISSION_TYPES)).stream().anyMatch(permission.permissionType::equalsIgnoreCase)) {
+            throw new InvalidPermissionException(
+                "The permission type for '"
+                    + permission.permissionString
+                    + "' is not valid. Valid permission types are: CLUSTER, INDICES, PLUGIN, and EXTENSION."
+            );
+        }
+
+        // Require a valid resource pattern for permissions based on indices, plugins, or extensions
+        if (permission.permissionType.equalsIgnoreCase("INDICES")
+            || permission.permissionType.equalsIgnoreCase("PLUGIN")
+            || permission.permissionType.equalsIgnoreCase("EXTENSION")) {
+            if (permission.resource.isEmpty()) {
```

Comment:
Isn't the empty resource pattern equivalent to `*`?  If some permissions types are treated differently covert QUALIFIED_PERMISSION_TYPES to an enum type with a way to declare the differences, see an [example](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/auditlog/config/AuditConfig.java#L174) of complex enum declaration.

Reply:
Craig had suggested making this into a enum too. I didn't before because there is not a clean .contains() function for using with enums. You either have to use Apache utilities or iterate over the elements. That being said, I will look at your link and see if maybe there is a trick in there that can be used to swap things over.

Reply:
For your first question about the empty pattern, I thought that an empty resource patter would be invalid. I was under the impression that you needed to explicitly specify `*` to get all the resources of a type. On the documentation page it says: 
![image](https://user-images.githubusercontent.com/65832608/214864852-7b5a6fc9-8578-457b-8511-d3613d733435.png)

Which reads to me like you would need to explicitly state `*` in order to get all of the resources of a specific type. Maybe I am misinterpreting this.

Reply:
I made a complex enum that I _believe_ looks like it should. I looked up some documentation but am not an enum expert so you may have a comment or two.